### ISSUE TITLE
multi-gpu seems to be slower than single gpu

### ISSUE THREAD

**Issue by hyciswu** [uber/horovod#71]
I have tried to run on single and multiple-gpus and multi-gpu seems to be slower on horovod on the tensor_mnist.py example. Below is the outputs
Here is single GPU run with `CUDA_VISIBLE_DEVICE=1 python  tensorflow_mnist.py`
```bash
INFO:tensorflow:loss = 2.30667, step = 1
INFO:tensorflow:loss = 2.29006, step = 11 (0.068 sec)
INFO:tensorflow:loss = 2.25554, step = 21 (0.052 sec)
INFO:tensorflow:loss = 2.1974, step = 31 (0.052 sec)
INFO:tensorflow:loss = 1.89438, step = 41 (0.051 sec)
INFO:tensorflow:loss = 1.82758, step = 51 (0.052 sec)
INFO:tensorflow:loss = 0.915703, step = 61 (0.049 sec)
INFO:tensorflow:loss = 1.77014, step = 71 (0.050 sec)
INFO:tensorflow:loss = 1.03163, step = 81 (0.049 sec)
INFO:tensorflow:loss = 0.936019, step = 91 (0.050 sec)
INFO:tensorflow:global_step/sec: 162.7
INFO:tensorflow:loss = 0.522648, step = 101 (0.048 sec)
INFO:tensorflow:loss = 0.573508, step = 111 (0.049 sec)
INFO:tensorflow:loss = 0.214133, step = 121 (0.049 sec)
INFO:tensorflow:loss = 0.757816, step = 131 (0.048 sec)
INFO:tensorflow:loss = 0.190643, step = 141 (0.047 sec)
INFO:tensorflow:loss = 0.213472, step = 151 (0.048 sec)
INFO:tensorflow:loss = 0.0751245, step = 161 (0.048 sec)
INFO:tensorflow:loss = 0.115839, step = 171 (0.048 sec)
INFO:tensorflow:loss = 0.0470715, step = 181 (0.048 sec)
INFO:tensorflow:loss = 0.173935, step = 191 (0.048 sec)
INFO:tensorflow:global_step/sec: 207.711
INFO:tensorflow:loss = 0.196011, step = 201 (0.048 sec)
INFO:tensorflow:loss = 0.0540665, step = 211 (0.049 sec)
INFO:tensorflow:loss = 0.157532, step = 221 (0.047 sec)
INFO:tensorflow:loss = 0.392506, step = 231 (0.048 sec)
INFO:tensorflow:loss = 0.216696, step = 241 (0.048 sec)
INFO:tensorflow:loss = 0.146749, step = 251 (0.048 sec)
INFO:tensorflow:loss = 0.0601752, step = 261 (0.047 sec)
INFO:tensorflow:loss = 0.255786, step = 271 (0.049 sec)
INFO:tensorflow:loss = 0.116301, step = 281 (0.048 sec)
INFO:tensorflow:loss = 0.226139, step = 291 (0.048 sec)
```

here is cmd for 2 gpus `CUDA_VISIBLE_DEVICES=1,2 mpirun -np 2 python tensorflow_mnist.py `
and the outputs
```bash
INFO:tensorflow:loss = 2.29663, step = 2
INFO:tensorflow:loss = 2.30619, step = 2
INFO:tensorflow:loss = 2.28995, step = 12 (0.295 sec)
INFO:tensorflow:loss = 2.2678, step = 12 (0.208 sec)
INFO:tensorflow:loss = 2.26281, step = 22 (0.189 sec)
INFO:tensorflow:loss = 2.2539, step = 22 (0.189 sec)
INFO:tensorflow:loss = 2.17149, step = 32 (0.190 sec)
INFO:tensorflow:loss = 2.18344, step = 32 (0.190 sec)
INFO:tensorflow:loss = 1.73136, step = 42 (0.190 sec)
INFO:tensorflow:loss = 1.66357, step = 42 (0.190 sec)
INFO:tensorflow:loss = 1.18375, step = 52 (0.190 sec)
INFO:tensorflow:loss = 1.1508, step = 52 (0.190 sec)
INFO:tensorflow:loss = 0.689034, step = 62 (0.190 sec)
INFO:tensorflow:loss = 0.60923, step = 62 (0.190 sec)
INFO:tensorflow:loss = 0.417928, step = 72 (0.189 sec)
INFO:tensorflow:loss = 0.389533, step = 72 (0.189 sec)
INFO:tensorflow:loss = 1.40314, step = 82 (0.194 sec)
INFO:tensorflow:loss = 1.60605, step = 82 (0.194 sec)
INFO:tensorflow:loss = 0.517931, step = 92 (0.198 sec)
INFO:tensorflow:loss = 0.588845, step = 92 (0.198 sec)
INFO:tensorflow:loss = 0.958909, step = 102 (0.190 sec)
INFO:tensorflow:global_step/sec: 49.6269
INFO:tensorflow:loss = 1.07172, step = 102 (0.191 sec)
INFO:tensorflow:loss = 0.209173, step = 112 (0.189 sec)
INFO:tensorflow:loss = 0.172761, step = 112 (0.188 sec)
INFO:tensorflow:loss = 0.282635, step = 122 (0.192 sec)
INFO:tensorflow:loss = 0.316832, step = 122 (0.192 sec)
INFO:tensorflow:loss = 0.235262, step = 132 (0.192 sec)
INFO:tensorflow:loss = 0.39039, step = 132 (0.192 sec)
INFO:tensorflow:loss = 0.249453, step = 142 (0.190 sec)
INFO:tensorflow:loss = 0.220639, step = 142 (0.190 sec)
INFO:tensorflow:loss = 0.188635, step = 152 (0.191 sec)
INFO:tensorflow:loss = 0.237106, step = 152 (0.191 sec)
INFO:tensorflow:loss = 0.0990839, step = 162 (0.196 sec)
INFO:tensorflow:loss = 0.298068, step = 162 (0.196 sec)
INFO:tensorflow:loss = 0.289399, step = 172 (0.190 sec)
INFO:tensorflow:loss = 0.209013, step = 172 (0.190 sec)
INFO:tensorflow:loss = 0.187031, step = 182 (0.195 sec)
INFO:tensorflow:loss = 0.325866, step = 182 (0.195 sec)
INFO:tensorflow:loss = 0.106716, step = 192 (0.193 sec)
INFO:tensorflow:loss = 0.120047, step = 192 (0.193 sec)
INFO:tensorflow:global_step/sec: 51.9471
```

**Comment by alsrgv** [comment]
@hyciswu, MNIST examples are made to just demonstrate the API, not as a performance benchmark. See [this comment](https://github.com/uber/horovod/issues/60#issuecomment-340137721) for instructions about running larger-scale CNN benchmarks which show scaling.

**Comment by hyciswu** [comment]
here is what i did with a much bigger network on the mnist, with 100 conv layers, the scaling still doesn't seem well

for 1 GPU
```bash
INFO:tensorflow:loss = 2.30251, step = 5
INFO:tensorflow:loss = 2.30294, step = 15 (1.941 sec)
INFO:tensorflow:loss = 2.30128, step = 25 (1.798 sec)
INFO:tensorflow:loss = 2.30023, step = 35 (1.867 sec)
INFO:tensorflow:loss = 2.30339, step = 45 (1.819 sec)
INFO:tensorflow:loss = 2.30061, step = 55 (1.886 sec)
INFO:tensorflow:loss = 2.30383, step = 65 (1.841 sec)
INFO:tensorflow:loss = 2.30526, step = 75 (1.849 sec)
INFO:tensorflow:loss = 2.30201, step = 85 (1.870 sec)
INFO:tensorflow:loss = 2.29636, step = 95 (1.985 sec)
INFO:tensorflow:global_step/sec: 5.10516
INFO:tensorflow:loss = 2.30484, step = 105 (1.843 sec)
INFO:tensorflow:loss = 2.29963, step = 115 (2.098 sec)
INFO:tensorflow:loss = 2.30506, step = 125 (1.852 sec)
INFO:tensorflow:loss = 2.29644, step = 135 (1.827 sec)
INFO:tensorflow:loss = 2.30231, step = 145 (1.884 sec)
INFO:tensorflow:loss = 2.30311, step = 155 (1.837 sec)
INFO:tensorflow:loss = 2.30286, step = 165 (1.848 sec)
INFO:tensorflow:loss = 2.30623, step = 175 (1.785 sec)
INFO:tensorflow:loss = 2.29542, step = 185 (1.829 sec)
INFO:tensorflow:loss = 2.29051, step = 195 (1.876 sec)
INFO:tensorflow:global_step/sec: 5.3379
```

----
for 4 GPUS

```bash
INFO:tensorflow:loss = 2.30266, step = 4
INFO:tensorflow:loss = 2.3026, step = 4
INFO:tensorflow:loss = 2.30253, step = 4
INFO:tensorflow:loss = 2.30219, step = 14 (3.514 sec)
INFO:tensorflow:loss = 2.3022, step = 14 (3.364 sec)
INFO:tensorflow:loss = 2.30237, step = 14 (3.709 sec)
INFO:tensorflow:loss = 2.30258, step = 14 (2.249 sec)
INFO:tensorflow:loss = 2.30315, step = 24 (2.031 sec)
INFO:tensorflow:loss = 2.30094, step = 24 (2.029 sec)
INFO:tensorflow:loss = 2.30253, step = 24 (2.031 sec)
INFO:tensorflow:loss = 2.30315, step = 24 (2.032 sec)
INFO:tensorflow:loss = 2.30258, step = 34 (2.034 sec)
INFO:tensorflow:loss = 2.30133, step = 34 (2.034 sec)
INFO:tensorflow:loss = 2.30255, step = 34 (2.034 sec)
INFO:tensorflow:loss = 2.30311, step = 34 (2.030 sec)
INFO:tensorflow:loss = 2.30298, step = 44 (2.025 sec)
INFO:tensorflow:loss = 2.30162, step = 44 (2.025 sec)
INFO:tensorflow:loss = 2.30053, step = 44 (2.025 sec)
INFO:tensorflow:loss = 2.30209, step = 44 (2.027 sec)
INFO:tensorflow:loss = 2.29936, step = 54 (2.039 sec)
INFO:tensorflow:loss = 2.30162, step = 54 (2.039 sec)
INFO:tensorflow:loss = 2.30054, step = 54 (2.039 sec)
INFO:tensorflow:loss = 2.30204, step = 54 (2.038 sec)
INFO:tensorflow:loss = 2.29892, step = 64 (2.032 sec)
INFO:tensorflow:loss = 2.30227, step = 64 (2.032 sec)
INFO:tensorflow:loss = 2.30007, step = 64 (2.032 sec)
INFO:tensorflow:loss = 2.30029, step = 64 (2.031 sec)
INFO:tensorflow:loss = 2.30478, step = 74 (2.024 sec)
INFO:tensorflow:loss = 2.30463, step = 74 (2.026 sec)
INFO:tensorflow:loss = 2.29902, step = 74 (2.027 sec)
INFO:tensorflow:loss = 2.29807, step = 74 (2.027 sec)
INFO:tensorflow:loss = 2.29651, step = 84 (2.021 sec)
INFO:tensorflow:loss = 2.29841, step = 84 (2.020 sec)
INFO:tensorflow:loss = 2.30109, step = 84 (2.025 sec)
INFO:tensorflow:loss = 2.29532, step = 84 (2.020 sec)
INFO:tensorflow:loss = 2.29895, step = 94 (2.022 sec)
INFO:tensorflow:loss = 2.28654, step = 94 (2.020 sec)
INFO:tensorflow:loss = 2.29804, step = 94 (2.022 sec)
INFO:tensorflow:loss = 2.30167, step = 94 (2.022 sec)
INFO:tensorflow:loss = 2.2955, step = 104 (2.024 sec)
INFO:tensorflow:loss = 2.30839, step = 104 (2.024 sec)
INFO:tensorflow:loss = 2.30464, step = 104 (2.024 sec)
INFO:tensorflow:global_step/sec: 4.58787
```

**Comment by hyciswu** [comment]
I am looking at the `global_step/sec`, but i guess that doesn't really say anything about the scaling? I guess a fair comparison is the loss at different time.

**Comment by alsrgv** [comment]
@hyciswu, global_step/sec here is actually a number of **synchronous** steps among all workers. 

So, if you had batch size 64 on one GPU and your process ten batches in 1.876 sec, you process 64 * 10 / 1.876 = 341.15 images/sec. And if you have batch size 64 on each of 4 GPUs and you process ten batches in 2.024 sec, you process 64 * 4 * 10 / 2.024 = 1264.82 images/sec which is 3.71x faster.

**Comment by hyciswu** [comment]
@alsrgv thanks. Got it.

**Comment by vitalyli** [comment]
Hi Alex,

I just want to confirm my calc on this. Running on 4 machines with 8 GPUs each; batch size 5000.

Here are actual performance results:
INFO:tensorflow:global_step/sec: 18.4019
INFO:tensorflow:step = 17700, loss = 0.6280955 (5.430 sec)
INFO:tensorflow:step = 17800, loss = 0.6355628 (5.313 sec)
INFO:tensorflow:step = 17800, loss = 0.6375086 (5.313 sec)
INFO:tensorflow:step = 17800, loss = 0.6374152 (5.313 sec)
INFO:tensorflow:step = 17800, loss = 0.6369878 (5.315 sec)
INFO:tensorflow:step = 17800, loss = 0.6382105 (5.314 sec)

So it's running 18.4 Global steps per second, we are running on 32 GPUs, each step is a batch of 5000 samples. 

Thus it's doing 18.4x32*5000 → 2.944Mil samples per second or 10.6 Billion samples per hour.
